src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -416,19 +417,27 @@ DatabaseShardImp::importShard(
         return false;
     }
 
-    std::lock_guard lock(mutex_);
-    auto const it {shards_.find(shardIndex)};
-    if (it == shards_.end() ||
-        it->second.shard ||
-        it->second.state != ShardInfo::State::import)
     {
-        JLOG(j_.error()) <<
-            "shard " << shardIndex << " failed to import";
-        return false;
+        std::lock_guard lock(mutex_);
+        auto const it {shards_.find(shardIndex)};
+        if (it == shards_.end() ||
+            it->second.shard ||
+            it->second.state != ShardInfo::State::import ||
+            it->second.retryFinalize)
+        {
+            JLOG(j_.error()) <<
+                "shard " << shardIndex << " failed to import";
+            shard.reset();
+            renameDir(dstDir, srcDir);
+            return false;
+        }
+
+        it->second.shard = std::move(shard);
+        it->second.retryFinalize =
+            std::make_unique<RetryFinalize>();
     }
 
-    it->second.shard = std::move(shard);
-    finalizeShard(it->second, true, lock);
+    finalizeWithRefHash(shardIndex);
```

Comment:
Appreciate the thoughtful feedback. But I'm afraid I have to disagree with this suggestion.

> To me it feels like it should be the ShardArchiveHandler's obligation to make sure there's an expected hash available for the last ledger in the shard we're importing.

The `ShardArchiveHandler` is just that, an archive handler; not an archive verifier. Rather, it's the job of the `DatabaseShardImp` to leverage data stores in performing verification/validation.

>  If any retries happen, or if something gets stalled, that should be handled by the ShardArchiveHandler.

Failure to retrieve the last ledger hash for a shard does not stall the `DatabaseShardImp`, it merely prolongs the time the shard spends in an invalidated state. For this reason I see no harm in leaving the responsibility of performing the retrieval with the `DatabaseShardImp`.

> To me it feels like it should be the ShardArchiveHandler's obligation to make sure there's an expected hash available for the last ledger in the shard we're importing. If that expected hash is not available, then the shard shouldn't even be handed over to DatabaseShard for import. If any retries happen, or if something gets stalled, that should be handled by the ShardArchiveHandler.

I don't feel comfortable with extending the scope of the `ShardArchiveHandler`, a class that is intended to operate simply as a download manager for shard archives, by also requiring it to perform preliminary verification on the files its downloading. Lets keep all the validation in one place, namely the `DatabaseShardImp`. As it is, it seems I already need to address the fragility and ugliness of this class as you mentioned, so lets avoid complicating it further by creating this new requirement?

> 
> Here's how I envision it going.
> 
>     1. The `ShardArchiveHandler` imports the shard.
> 
>     2. The `ShardArchiveHandler` verifies that the expected hash of the imported shard is available.
>        
>        1. If the expected hash is _not_ available, the `ShardArchiveHandler` _stalls_ waiting for the hash to become available.
>        2. If the hash doesn't become available after a reasonable time then the shard is discarded and the `ShardArchiveHandler` moves on to the next shard.  If there are no more shards to import it stops.
> 
>     3. Since the expected hash is available, the `ShardArchiveHandler` calls `DatabaseShard::importShard()`.
> 
>     4. If (due to a bug) `DatabaseShardImp::importShard()` is unable to get the expected hash of the imported shard, then the imported shard is thrown out immediately.  The shard is not even added to the shard store.
> 
>     5. Otherwise the expected hash _is_ available.  `DatabaseShardImp::importShard()` immediately proceeds with the full verification.  If the verification does not pass the shard is thrown out.
> 
>     6. The verification succeeded, so the shard is immediately added to the database.


The problem I have with this is the inefficiency in retrieving the expected hash twice, and stalling the progress of the `ShardArchiveHandler` when it should be moving onto the next download.

> And I don't think that the DatabaseShardImp should have any responsibility for waiting around on any provisionally ready shards.

This is already an existing feature of the DatabaseShardImp as it holds shards that are in the process of being validated. Maybe you feel that this class should not store pending shards - perhaps by creating some auxillary class to facilitate this. If that's the case I think that deserves a larger discussion as part of a separate project. I'm reluctant to introduce that scope creep into this PR.

Again, remember that the introduction of this expected hash verification is not introducing a 0 -> 1 change in the delay of imported shards. The shard import will be delayed no matter what, since the validation of the shard is a lengthy operation.

Reply:
Sure.  Fair enough.  You and @miguelportilla need to have a design that you're both comfortable with.  If the `DatabaseShardImp` already has the feature of handling unverified shards, adding yet another form of unverified shard makes some sense.

However consider that with this change a class named `DatabaseShardImp` now has a _timer_ in it that re-checks whether enough information is available to validate something that it has _already_ stored.  My gut reaction is that a timer probably doesn't belong in something called a `DatabaseImp`.  To me it feels like that work is being done in the wrong place.  If I were to choose where the timer belongs, it feels more natural in the `ShardArchiveHandler` given the two choices I see.

And also, for me, I'd rather verify something _before_ I put it in the database than to 1) put it in the database, then 2) verify it, then 3) throw it out if the validation fails.

Occasions where the `ShardArchiveHandler` is stalled because it has a shard that can't be verified yet will be rare.  It only happens when importing a shard that is ahead of what the server has validated.  So the throughput hit to the `ShardArchiveHandler` will be small.  And that throughput hit could help keep some complexity out of the `ShardDatabaseImp`.

Maybe another approach would be to not even allow the import of a shard to start unless the local validated ledger is ahead of the requested shard.  Then a shard that is ahead of the local validated ledger should never show up.  If one were to show up, then it's just an error and you should throw the (currently) unverifiable shard away.  That might be an approach that simply removes the timer.

I'm not insisting that this is a change you and @miguelportilla must make.  But the current design feels out-of-balance to me.  My approach to how to solve the out-of-balance may be the wrong one.  I may also just have bad instincts on this one.  But you have not yet convinced me that you've made the right design trade off here.

Reply:
@scottschurr You've brought up some valid points, thanks!

I agree that the added timer and related retry logic in `DatabaseShardImp` doesn't seem like a good design fit as is. `DatabaseShardImp` has existing functionality that queues work on shards in separate threads. This is a necessary part of the process to finalize acquired or imported shards. But it doesn't mean the timer retry logic should use it. Perhaps moving the timer/retry portion of the code into  `ShardArchiveHandler` is a good idea.

> And also, for me, I'd rather verify something before I put it in the database than to 1) put it in the database, then 2) verify it, then 3) throw it out if the validation fails.

I think the shard store exceeds the responsibility of a regular database. Internally it utilizes many (can be thousands) relational and key value databases through shards but it also manages other things like notifying peers of status updates. It inherited the database name from its need to be compatible with the node store and its base class, but that will eventually change. I think of the shard store as more of a service that manages many databases.

Verifying the NuDB portion of a shard could be externalized but there is a good reason we don't. In order to validate a shard's NuDB database, we need to walk its SHAMaps and verify the Merkle trees. As you can imagine, this is very time consuming and a storage intensive operation. Our discussion mentions `validate` but there is much more involved. Whether acquired or imported, shards must go through a `finalize` process. During this phase we validate data but also generate and optimize SQLite databases. This must be done as we walk the SHAmaps. NuDB databases will also need to be re-written to be deterministic files. An external validation would cause us to visit the data twice, which is prohibitively expensive. A shard that is being finalized isn't available or considered `stored` until the `finalize`  process successfully completes.

I hope this explanation helps a bit.

Reply:
@miguelportilla, yes thanks, that explanation helps a great deal.  I now see that the `ShardDatabaseImp` is a lot more than a database.  And, yes, it's important that the `validate` and `finalize` processes be efficient.  So your explanation helps me see why the `ShardDatabaseImp` has the various states for the databases that it shepherds.  Thanks.

My intention was to suggest that the `ShardArchiveHandler` not surrender the imported shard to the `ShardDatabaseImp` until such time that all necessary information is available to validate the shard.  I think that required information is 1) the shard's contents plus 2) the expected hash of the last ledger in the shard.

If the `ShardDatabaseImp` can be assured that when it receives an imported shard the expected hash of the last ledger in the shard is immediately locally available, then the `ShardDatabaseImp` doesn't need the timer; it can put the validation process in the `JobQueue` immediately.  I think that could be accomplished by:

1. The `ShardArchiveHandler` having the timer or, simpler,
2. The `ShardArchiveHandler` not even requesting the shard import from the network unless it knows the the expected hash of the last ledger of the shard is available locally.

I was initially suggesting approach 1, but I now think option 2 is much better.  There may be good reasons why these (or other approaches) are not viable options.  But I think they are at least worth considering.  Thanks for your patience.

Reply:
@scottschurr I like your suggestion of not surrendering the imported shard until we have all of the information. That would move the timer/retry complexity from the shard store code into the archiveHandler. 

On Solution 2, I would expect the RPC return message to explain which of the shards indexes specified in the command would not be downloaded and why. That should be quite doable without much fuss. But this solution makes life a bit more difficult for the server operator as they would need to retry failed shards after they have confirm (through polling?) their server has acquired newer history. Solution 1 does that work for the operator.

Reply:
@miguelportilla, thanks for considering these options.

In terms of error reporting, it seems like both of these approaches have the same error reporting problem.  Using solution 1 if the necessary information is not available in five minutes, then the server operator needs to be told about the failure.   However it is true that solution 1 will present these errors to the user less often.

My suspicion is that these sorts of errors will be rare using either approach.  But I don't have actual user experience to back that up.

The five minute delay on error reporting in solution 1 has an upside and a downside.  Consider these failing cases:

**Solution 1:** The server operator specifies a shard to import.  Five minutes later the server returns an error that a certain shard could not be imported.  Where does that error show up?  In the log?  Will the server operator see it?

**Solution 2:** The server operator specifies a shard to import.  The server sees that the required ledger hash is not available and can tell the operator immediate that the operation failed and should be retried when the hash is available.  The error response is immediate which improves the likelihood the error is seen and improves the operator's comprehension.

It's true that the error will happen less often with solution 1.  But when solution 1 has an  error it is more confusing.

Just thoughts to consider...

Reply:
@scottschurr thanks for the feedback. I discussed with @miguelportilla and while either approach brings its own benefits and drawbacks, I'm going to go with _solution 1_.

With this approach failures are less likely, meaning the user won't have to manually retry as frequently. Additionally, by design, if the reason for the failure is due to the server being out of sync, the pending shards are held until the server re-syncs to the network.

We anticipate this command will be used to download large batches of shards, so one can imagine a scenario with _solution 2_ wherein twenty shards are able to download immediately, but eighty fail. Since the user gets feedback immediately, they might be encouraged to retry right away, forcing the user to parse the json response and reconstruct the download request which could be a hassle. In any case, the immediate retry will also fail, since the requests won't be accepted until the current set of downloads have completed, meaning the user now has to choose between cancelling the current set and retrying the full batch again (to avoid parsing the failures), or wait until the current set has completed and then retry with the set of failures (forcing the user to wait around for the downloads to complete if he wishes to complete all downloads as quickly as possible). With _solution 1_, attempts made to verify the hash while the server is not in sync are not counted toward the attempt limit, increasing the chance of successfully completing downloads, and making the user's job just a bit easier.

So in summary, I'm opting for an easier user experience and fewer failures at the cost of delayed and passive error reporting.

Reply:
I'm good with your choice, but I'm surprised when you say that 80 requested shards could fail.  Doesn't the failure only occur if the expected ledger hash of the last ledger in the shard is not available?  If that's correct then I think we should only see failures...

1. If they are requesting a very recently published shard and the local machine has not quite finished processing the ledger (in this case the hash should be available in a few seconds),

2. The local machine is not in sync with the network, hashes of recent ledgers are not locally available (in which case we need to tell them to get their machine in sync), or

3. They are asking for a shard from the future, and maybe that shouldn't work.

But I could very easily be getting my story confused.

Reply:
Unless I'm mistaken, this can happen if the server has been synced, but has not yet built sufficient history to satisfy all of the necessary last ledger hashes.

As for the implementation of the hash confirmation retries, I'm considering two approaches:

1. After a download completes, the `ShardArchiveHandler` halts further downloads until the completed shard either has its hash verified or is discarded.

2. After a download completes, the `ShardArchiveHandler` hands off the shard to an auxiliary class (owned by the handler) that queues pending shards and attempts hash confirmation one at a time.

I'm leaning toward the second approach because it allows us to complete downloads more quickly, but it is more complicated. Now we have two containers for keeping track of shards, and the state database will need to be modified to track archives that are in the pending queue.

Any thoughts?

Reply:
So not being able to get the ledger hash happens with _old_ shards that the local server does not have enough history to have the ledger hashes in the skip list.  Okay, that makes sense.  Thanks for the explanation.  When that situation occurs, what does the `ShardArchiveHandler` do to try and get the expected hash from the network?  I think I dropped that particular ball in my code review.

Regarding these two approaches, I don't view getting shards stored locally as something we need to be in a huge hurry about.  Choosing between a relatively simple, robust solution that takes a couple of minutes longer that a highly optimized approach, I'd prefer the simple robust solution.  It's easier to code review.  It probably has fewer bugs.  It still gets the job done.

But, again, maybe there is a good reason (that I'm not aware of) to try and get the shard import done in a big hurry...

Reply:
As long as there is a validated ledger whose index is greater than the index of the expected ledger, the attempt will count, and after five attempts, the shard is discarded.

I agree with going with the simple option in this case. I don't see a good reason for rushing the download process and complicated the handler. I'll go with approach 1.

Reply:
Regarding the needed ledger hash, naively, it seems that unless you tell the server to do something active the needed ledger hash for an _old_ shard is unlikely to just show up spontaneously over a five minute interval.  I can definitely see how this would work for a brand new shard assuming the server is synced (although it should take a lot less than five minutes).

Still feels like I'm missing something important on this topic...

I feel good about choosing the simpler solution for shard import.  Thanks for that.

Reply:
Well, the idea is that if we're still retrieving old history, the heuristic that checks the sequence of the most recently validated ledger (_solution 2_) won't return true, and so these attempts made while still waiting for old stuff won't count toward the max attempts. But with _solution 1_, we don't have any sort of check, we just fail immediately if the expected hash isn't available right away. So those shards whose expected hashes happen to be available ATM will pass and everything else fails.

Reply:
Hmmm.  I'm really sorry.  I don't mean to be stretching out the discussion.  I really want to get out of the way and let you do your work.  You've been outstandingly accommodating.  Thanks so much for your patience.

>  if we're still retrieving old history, the heuristic that checks the sequence of the most recently validated ledger (solution 2) won't return true...

If we're retrieving _old_ shards, then the check to see if the ledger sequence of the last ledger in the old shard is less than the most recently validated ledger _will_ return true, right?  I feel like I've dropped something again.

I'm extra sorry now, because I'm going to beat a dead horse.  When I compare...

1. the cost of making sure I have the expected hash of a ledger in local storage vs.
2. the cost of downloading a shard (lots of ledgers worth) over the network

it feels like downloading the shard is way more work for everybody, both locally and for whoever serves the shard.  Especially if that downloaded shard is then thrown away because we later find out that we didn't have the ledger hash necessary for validation.  So, to me, it feels like it makes sense to verify that exactly the necessary ledger hash is available before we even start importing the shard.  The `ShardArchiveHandler` doesn't need to use that hash, it should just verify that it's available.

I think it's really the same work, just in a different order.  If you want to wait up to five minutes, that's cool.  But wait up to five minutes for the required ledger hash to show up _first_.  If it does show up, then start the shard download.  If it doesn't show up then kill the import (with an error message) without even requesting the shard, much less downloading it.

But I've said that before, and it doesn't make sense to you, and you need to do what feels right to you.  Sorry for bringing that up again.  No response necessary.

Reply:
> If we're retrieving old shards, then the check to see if the ledger sequence of the last ledger in the old shard is less than the most recently validated ledger will return true, right? I feel like I've dropped something again.

My understanding that this wasn't the case, but maybe you're right. I thought that a recently validated ledger wouldn't be available while history is still incomplete. Again, this isn't my heuristic, and I haven't yet had the time to research all the mechanics involved in this. But since this is just to explain one hypothetical scenario with multiple errors with one of the proposed solutions, I hope its ok that we can discuss that later.

> I think it's really the same work, just in a different order. If you want to wait up to five minutes, that's cool. But wait up to five minutes for the required ledger hash to show up first. If it does show up, then start the shard download. If it doesn't show up then kill the import (with an error message) without even requesting the shard, much less downloading it.

> But I've said that before, and it doesn't make sense to you, and you need to do what feels right to you. Sorry for bringing that up again. No response necessary.

Yes, this is the approach I'm planning on using, I just got the order wrong in explaining the details. Sorry for the confusion.